angular/projects/global/src/lib/components/table-template/table-template-utils.ts

```diff
@@ -0,0 +1,87 @@
+import { Injectable } from '@angular/core';
+import { Router, Params } from '@angular/router';
+import { PlatformLocation } from '@angular/common';
+import {
+  TableObject,
+  DEFAULT_TABLE_PAGE_SIZE,
+  DEFAULT_TABLE_CURRENT_PAGE,
+  DEFAULT_TABLE_SORT_BY,
+  DEFAULT_TABLE_KEYWORDS
+} from './table-object';
+
+@Injectable()
+export class TableTemplateUtils {
+  constructor(private platformLocation: PlatformLocation, private router: Router) {}
+
+  public updateUrl(tableObject: TableObject) {
+    let currentUrl = this.router.url;
+    currentUrl = (this.platformLocation as any).getBaseHrefFromDOM() + currentUrl.slice(1);
+    currentUrl = currentUrl.split(';')[0];
+    currentUrl += `;currentPage=${tableObject.currentPage};pageSize=${tableObject.pageSize}`;
+    if (tableObject.keywords !== '') {
+      currentUrl += `;keywords=${tableObject.keywords}`;
+    }
+    if (tableObject.sortBy !== '' && tableObject.sortBy !== null) {
+      currentUrl += `;sortBy=${tableObject.sortBy}`;
+    }
+    if (tableObject.filter !== null && tableObject.filter !== {}) {
+      Object.keys(tableObject.filter).forEach(key => {
+        if (tableObject.filter[key] === true || tableObject.filter[key] === false) {
+          currentUrl += `;${key}=${tableObject.filter[key]}`;
+        } else {
+          currentUrl += `;${key}=`;
+          tableObject.filter[key].split(',').forEach(item => {
+            currentUrl += `${item},`;
+          });
+          currentUrl = currentUrl.slice(0, -1);
+        }
+      });
+    }
+    currentUrl += ';ms=' + new Date().getTime();
+    window.history.replaceState({}, '', currentUrl);
+  }
+
+  public updateTableObjectWithUrlParams(
+    routeParams: Params,
+    tableObject: TableObject,
+    forceUpdateUrl = false,
+    filterFieldList = []
+  ) {
+    tableObject.pageSize = routeParams.pageSize ? +routeParams.pageSize : DEFAULT_TABLE_PAGE_SIZE;
+    // tslint:disable-next-line: max-line-length
+    tableObject.currentPage = routeParams.currentPage ? +routeParams.currentPage : DEFAULT_TABLE_CURRENT_PAGE;
+    tableObject.sortBy = routeParams.sortBy ? routeParams.sortBy : DEFAULT_TABLE_SORT_BY;
+    tableObject.keywords = routeParams.keywords ? routeParams.keywords : DEFAULT_TABLE_KEYWORDS;
+    filterFieldList.map(field => {
+      if (routeParams[field]) {
+        tableObject.filter[field] = routeParams[field];
+      }
+    });
+
+    // TODO: Get fields from URL.
+
+    if (forceUpdateUrl) {
+      this.updateUrl(tableObject);
```

Comment:
This looks like its purely for convenience, so someone can update the table object and the url in one call, rather than having to call both.

This is just my preference, but I like it when this sort of thing doesnt exist.  If the consumer makes 2 calls (updateUrl(), updateTableObject()) its really clear what that does.  But if you just see (updateTableObjct(true)) its not as clear immediately what that is doing.  

It also becomes harder to make updates to these functions, because you need to track down who is calling it with true, and who with false, in order to know what other code you might affect.  Whereas if this function only does 1 thing, then its obvious what the affect will be.

It also technically makes your testing slightly more complex, because now you need to test the path where forceUpdateUrl is true and false, on top of everything else this function does.